### DISCUSSION TITLE
Parameter doubts

### DISCUSSION THREAD

**Discussion by Tianfneg** [TinyMPC/TinyMPC#36] [Q&A]
![image](https://github.com/TinyMPC/TinyMPC/assets/147623854/a4b3516c-2040-414c-80a8-697f1b5a90d2)
Hi, I would like to ask if this reference trajectory gives the absolute trajectory or the error between the reference trajectory and the real trajectory? Thanks for your help.

**Comment by xkhainguyen** [comment]
Hi,

The reference trajectory depends on your problem formulation. For example in the quadrotor case, it was relative to the hovering state around which we linearized the nonlinear dynamics model to obtain a LTI model.

**Reply by Tianfneg** [reply]
![image](https://github.com/TinyMPC/TinyMPC/assets/147623854/36368be7-4bae-40a1-9c8a-7c2642d89019)
![image](https://github.com/TinyMPC/TinyMPC/assets/147623854/80155576-4bfb-44f8-9ef3-2f64fa2ce064)

Q1：Hi, theoretically, the cost function should be an error in the form of an error solving for a minimum, but what you entered above seems to be a trajectory. You also don't seem to have treated it as a form of error in your linear project update.

Q2：Can you explain the rationale for updating the linear term so much? In the references I've seen, LQR doesn't have a linear term, so if you have any information, please send me an e-mail. wangtianfeng1107@gmail.com
Thank you very much and good luck with your research! 

> 你好
> 
> 参考轨迹取决于您的问题表述。例如，在四旋翼飞行器的情况下，它是相对于悬停状态而言的，我们围绕悬停状态对非线性动力学模型进行了线性化，从而获得了LTI模型。

**Reply by xkhainguyen** [reply]
Hi,

1. Yes the cost penalizes the error between horizon predicted vs reference trajectory, something like this: 
$\text{min} \sum (x - \bar x)^T Q (x - \bar x) = \text{min}\sum x^T Q x - 2 \bar x^T Q x =  \text{min}\sum x^T Q x + q^T x$
where we define $q^T = 2 \bar x^T Q $
We project updated $x$ directly into the feasible domain.

2. As above, you can see why the linear cost term $q$ appears. It depends on the reference trajectory within the current horizon (actually you can just shift it and compute the last index). This is standard general LQR formulation, you can look at https://underactuated.mit.edu/lqr.html#section5

**Reply by Tianfneg** [reply]
Good day, although I understood this part of the principle when I read your reply, your reply verified my inference. Thank you for solving every question I have had for such a long time your team, and finally I wish you all the best for your health and research!
